bots: Add github-push tool

mvollmer opened the pull request.

mvollmer commented:
Pull request has been rewritten.
Old commits: https://github.com/cockpit-project/cockpit/commits/6286bd30b97fd074780deb80255f26fd548150af
Diff:
```[diff]
diff --git a/bots/github-push b/bots/github-push
index a82e39f99..078bca340 100755
--- a/bots/github-push
+++ b/bots/github-push
@@ -62,10 +62,21 @@ def main():
         sys.stderr.write("{0} is not a pull request.\n".format(opts.pull))
         return 1
 
-    if git("merge-base", local, remote) == remote:
-        sys.stderr.write("Incremental change, fast forwarding.\n")
+    github_branch = pull["head"]["ref"]
+    github_remote = "git@github.com:%s.git" % pull["head"]["repo"]["full_name"]
+
+    if git("merge-base", local, remote).strip() == remote:
+        sys.stderr.write("Incremental change, fast forwarding\n")
+        git("push", github_remote, local + ":" + github_branch)
         return 0
 
+    if git("diff", "--status", remote, local) == "":
+        sys.stderr.write("Commit history rewrite\n")
+        # copy test results, simple comment
+        pass
+    else:
+        # big comment
+        pass
 
 if __name__ == '__main__':
     sys.exit(main())

```

mvollmer commented:
Pull request has been rewritten. [Old commits](https://github.com/cockpit-project/cockpit/commits/6286bd30b97fd074780deb80255f26fd548150af) [Diff](https://github.com/cockpit-project/cockpit/compare/6286bd30b97fd074780deb80255f26fd548150af..1e73f850e81e66b6abb403964c649040cd55808a)

mvollmer commented:
Pull request has been rewritten. [Old commits](https://github.com/cockpit-project/cockpit/commits/6286bd30b97fd074780deb80255f26fd548150af). [Diff](https://github.com/cockpit-project/cockpit/compare/6286bd30b97fd074780deb80255f26fd548150af..1e73f850e81e66b6abb403964c649040cd55808a).

mvollmer commented:
Pull request has been rewritten.
[Old commits](https://github.com/cockpit-project/cockpit/commits/0c5f32a7d34cbcfc6810390e6a27f5a415940a56) | [Diff](https://github.com/cockpit-project/cockpit/compare/0c5f32a7d34cbcfc6810390e6a27f5a415940a56..cbd3843a73b431c657a0710a2c5a76420056a180)
```diff
diff --git a/bots/github-push b/bots/github-push
index 18299dbc9..ad34c6a0b 100755
--- a/bots/github-push
+++ b/bots/github-push
@@ -106,9 +106,12 @@ def main():
 
     if diff == "":
         old_statuses = api.statuses(remote)
+        copy_count = 0
         for key in old_statuses:
             if old_statuses[key]["state"] != "pending":
                 api.post("statuses/" + local, old_statuses[key])
+                copy_count += 1
+        comment += "\n%s test results have been copied.\n" % copy_count
     else:
         comment += " | [Diff](%s)" % diff_link
         if diff.count("\n") < 50:
```

mvollmer commented:
Pull request has been rewritten.
[Old commits](https://github.com/cockpit-project/cockpit/commits/cbd3843a73b431c657a0710a2c5a76420056a180) | [Diff](https://github.com/cockpit-project/cockpit/compare/cbd3843a73b431c657a0710a2c5a76420056a180..e8bf0216319fa0d7beadecf9d480b22def2de5f1)

mvollmer commented:
Pull request has been rewritten.
[Old commits](https://github.com/cockpit-project/cockpit/commits/e8bf0216319fa0d7beadecf9d480b22def2de5f1)
0 test results have been copied.

mvollmer commented:
Commit history has been rewritten
[Old commits](https://github.com/cockpit-project/cockpit/commits/adf4f06ef984595ef70275a9b7ecea8b93b397ec)
26 test results have been copied.

mvollmer commented:
Pull request has been rewritten
[Old commits](https://github.com/cockpit-project/cockpit/commits/b11786cc966f6b6206c7c23e1dfe8386f77bf284) | [Diff](https://github.com/cockpit-project/cockpit/compare/b11786cc966f6b6206c7c23e1dfe8386f77bf284..7c10bf885974a5c8924331585dd0ad0fe2727d4a)
```diff
diff --git a/bots/github-push b/bots/github-push
index b5e11e2f8..4e4d4698d 100755
--- a/bots/github-push
+++ b/bots/github-push
@@ -109,7 +109,7 @@ def main():
             if old_statuses[key]["state"] != "pending":
                 api.post("statuses/" + local, old_statuses[key])
                 copy_count += 1
-        comment = "Commit history has been rewritten\n[Old commits](%s)" % old_commits_link
+        comment = "Commit history has been rewritten.\n[Old commits](%s)" % old_commits_link
         comment += "\n%s test results have been copied.\n" % copy_count
     else:
         comment = "Pull request has been rewritten\n[Old commits](%s)" % old_commits_link
```

mvollmer commented:
Pull request has been rewritten
[Old commits](https://github.com/cockpit-project/cockpit/commits/7c10bf885974a5c8924331585dd0ad0fe2727d4a) | [Diff](https://github.com/cockpit-project/cockpit/compare/7c10bf885974a5c8924331585dd0ad0fe2727d4a..a42cabe67f4fd082caeaa6e7bd260ab315a2482d)
```diff
diff --git a/bots/github-push b/bots/github-push
index 4e4d4698d..9047a2a11 100755
--- a/bots/github-push
+++ b/bots/github-push
@@ -22,10 +22,9 @@
 # The idea is that by using this tool, most (if not all) downsides of
 # force pushing to a PR go away.
 #
-# TODO - In this version, the tool expects that the local branch has a
-# upstream tracking branch configured, and that this tracking
-# branch corresponds to the branch of the pull request.  We can relax
-# that.
+# TODO - The tool expects that the local branch has a upstream
+# tracking branch configured, and that this tracking branch
+# corresponds to the branch of the pull request.  We can relax that.
 
 import subprocess
 import argparse
```

mvollmer commented:
Commit history has been rewritten.
[Old commits](https://github.com/cockpit-project/cockpit/commits/b90e7e52e62404eadd17dfd2bb2cb38c3fdd3b25)
25 test results have been copied.

mvollmer commented:
Pull request has been rewritten
[Old commits](https://github.com/cockpit-project/cockpit/commits/3dbb4da6b75988b79f1a7fd0c18b37bb0b026f7a) | [Diff](https://github.com/cockpit-project/cockpit/compare/3dbb4da6b75988b79f1a7fd0c18b37bb0b026f7a..df4a426e9249ee76aec5fb037fe62db708d2bef1)
```diff
diff --git a/bots/github-push b/bots/github-push
index e6b4703a3..9f9559b01 100755
--- a/bots/github-push
+++ b/bots/github-push
@@ -55,6 +55,7 @@
 import subprocess
 import argparse
 import sys
+import time
 
 from task import github
 
```

mvollmer commented:
New commits. [Diff](https://github.com/cockpit-project/cockpit/compare/df4a426e9249ee76aec5fb037fe62db708d2bef1..7301d4926210992b1a14989c3f755875c7052ea7)
```diff
diff --git a/bots/github-push b/bots/github-push
index 9f9559b01..a78ea9f7f 100755
--- a/bots/github-push
+++ b/bots/github-push
@@ -97,10 +97,6 @@ def main():
     push_refspecs = local + ":" + tracking_ref
     push_lease = tracking_ref + ":" + remote
 
-    if git("merge-base", local, remote).strip() == remote:
-        git("push", push_dest, push_refspecs)
-        return 0
-
     pull_number = None
     pull_config = git_config("pull.%s.%s" % (tracking_remote, tracking_ref))
     if not pull_config:
@@ -136,7 +132,12 @@ def main():
             print("(new commits not yet in the GiHub API...please stand by. *beep*)")
         time.sleep(1)
 
-    if diff == "":
+    if git("merge-base", local, remote).strip() == remote:
+        comment = "New commits. [Diff](%s)" % diff_link
+        if diff.count("\n") < 50:
+            comment += "\n```diff\n" + diff + "```"
+
+    elif diff == "":
         old_statuses = api.statuses(remote)
         copy_count = 0
         for key in old_statuses:
```

mvollmer commented:
New commits. [Diff](https://github.com/cockpit-project/cockpit/compare/7301d4926210992b1a14989c3f755875c7052ea7..537114d1fe16afba6ff07cb4cf9a9a5971baa951)

mvollmer commented:
New commits. [Diff](https://github.com/cockpit-project/cockpit/compare/537114d1fe16afba6ff07cb4cf9a9a5971baa951..43f6b196f8042d953ac52305598906aa21562e7a)
<details><summary>Changes</summary>

```diff
diff --git a/bots/github-push b/bots/github-push
index 114250249..a79623b97 100755
--- a/bots/github-push
+++ b/bots/github-push
@@ -79,7 +79,7 @@ def main():
     old_commits_link = "https://github.com/" + opts.repo + "/commits/" + remote
     diff_link = "https://github.com/" + opts.repo + "/compare/" + remote + ".." + local
     diff_text = ""
-    if diff.count("\n") < 50:
+    if diff.count("\n") < 200:
         diff_text = "\n<details><summary>Changes</summary>\n\n```diff\n" + diff + "```\n</details>"
 
     git("push", "--force-with-lease=" + push_lease, push_dest, push_refspecs)
```
</details>

mvollmer commented:
<details ><summary>New commits.</summary>
[Diff](https://github.com/cockpit-project/cockpit/compare/43f6b196f8042d953ac52305598906aa21562e7a..c7aaddca3298ce151e5a363d831476891a695ee4)

```diff
diff --git a/bots/github-push b/bots/github-push
index a79623b97..ae21ff427 100755
--- a/bots/github-push
+++ b/bots/github-push
@@ -75,12 +75,21 @@ def main():
         pull_number=int(pull_config)
 
     diff = git("diff",remote, local)
-
     old_commits_link = "https://github.com/" + opts.repo + "/commits/" + remote
     diff_link = "https://github.com/" + opts.repo + "/compare/" + remote + ".." + local
-    diff_text = ""
-    if diff.count("\n") < 200:
-        diff_text = "\n<details><summary>Changes</summary>\n\n```diff\n" + diff + "```\n</details>"
+
+    def make_comment(title, include_old_commits=False):
+        diff_lines = diff.count("\n")
+        if include_old_commits:
+            first_line = "[Old commits](%s) | [Diff](%s)" % (old_commits_link, diff_link)
+        else:
+            first_line = "[Diff](%s)" % (diff_link)
+        if diff_lines < 500:
+            diff_open = "open" if diff_lines < 50 else ""
+            return ("<details %s><summary>%s</summary>\n%s\n\n```diff\n%s```\n</details>\n"
+                    % (diff_open, title, first_line, diff))
+        else:
+            return "%s\n%s\n" % (title, first_line)
 
     git("push", "--force-with-lease=" + push_lease, push_dest, push_refspecs)
 
@@ -95,8 +104,7 @@ def main():
         time.sleep(1)
 
     if git("merge-base", local, remote).strip() == remote:
-        comment = "New commits. [Diff](%s)" % diff_link
-        comment += diff_text
+        comment = make_comment("New commits.")
 
     elif diff == "":
         old_statuses = api.statuses(remote)
@@ -105,14 +113,12 @@ def main():
             if old_statuses[key]["state"] != "pending":
                 api.post("statuses/" + local, old_statuses[key])
                 copy_count += 1
-        comment = "Commit history has been rewritten.\n[Old commits](%s)" % old_commits_link
+
+        comment = make_comment("Commit history has been rewritten.", include_old_commits=True)
         comment += "\n%s test results have been copied.\n" % copy_count
     else:
-        comment = "Pull request has been rewritten\n[Old commits](%s)" % old_commits_link
-        comment += " | [Diff](%s)" % diff_link
-        comment += diff_text
+        comment = make_comment("Pull request has been rewritten.")
 
-    comment += "\n"
     api.post("issues/{0}/comments".format(pull_number), { "body": comment })
 
 if __name__ == '__main__':
```
</details>

martinpitt commented on bots/github-push:
```diff
@@ -0,0 +1,125 @@
+        sys.stderr.write("Nothing to push.\n")
+        return 0
+
+    ( tracking_remote, tracking_ref ) = git("rev-parse", "--abbrev-ref", "--symbolic-full-name", "@{u}").strip().split("/")
```
Just to understand this: This is supposed to result in something like `(myremotename, branchname)`, right? If I run this on my `master` checkout, I get `(origin, master)`. (As said above, it doesn't work for me on branches)

mvollmer replied:
Can you show the relevant section from your `.git/config`?  Mine looks like this:

```
[branch "bots-github-push-tool"]
	remote = mvollmer
	merge = refs/heads/bots-github-push-tool
```

mvollmer replied:
> This is supposed to result in something like (myremotename, branchname), right? 

Yes, with the config above, I get `mvollmer/bots-github-push-tool` for both @{u} and  @{push}.

martinpitt commented on bots/github-push:
```diff
@@ -0,0 +1,125 @@
+            return 1
+        git("config", "pull.%s.%s" % (tracking_remote, tracking_ref), str(pull_number))
+    else:
+        pull_number=int(pull_config)
```
It seems worth moving the determination of the pull number into a function, with some comments.

mvollmer replied:
The code has been refactored quite a lot now. How does it look now?

martinpitt commented on bots/github-push:
````diff
@@ -0,0 +1,125 @@
+        if diff_lines < 500:
+            diff_open = "open" if diff_lines < 50 else ""
+            return ("<details %s><summary>%s</summary>\n\n%s\n```diff\n%s```\n</details>\n"
+                    % (diff_open, title, first_line, diff))
````
Is this magic GitHub syntax to create an expander in a comment? Nice!

mvollmer replied:
It's actually non-magic HTML... :-)

martinpitt commented on bots/github-push:
```diff
@@ -0,0 +1,125 @@
+            first_line = "[Old commits](%s) | [Diff](%s)" % (old_commits_link, diff_link)
+        else:
+            first_line = "[Diff](%s)" % (diff_link)
+        if diff_lines < 500:
```
Is there still any possibility to get larger diffs? It sounds like the `diff_link` above should already provide that? If so, why can't we just always use that? It seems more useful to me to go to a GitHub-provided diff page, as you can then also inline-comment on it, rather than a static comment.

mvollmer replied:
> Is there still any possibility to get larger diffs? It sounds like the diff_link above should already provide that? 

Yes.

> If so, why can't we just always use that? It seems more useful to me to go to a GitHub-provided diff page, as you can then also inline-comment on it, rather than a static comment.

That's a good point.  I thought inline diffs are nice because you don't need to jump between pages, you can just read them as you scroll down the page.  But maybe they are distracting... 

What about moving the link out of the expander and keeping the expander always closed initially?

martinpitt replied:
That seems fine. I also don't want to dwell on this too much -- we can always change it later if it becomes to clumsy or so; I was mostly curious how it works, and that it also supports large diffs.

martinpitt commented on bots/task/github.py:
```diff
@@ -167,7 +167,7 @@ def request(self, method, resource, data="", headers=None):
         }
 
-    def get(self, resource):
+    def get(self, resource, verbose = True):
```
Please no spaces around the `=` in default arguments.

martinpitt commented:
Also, I think this PR should remove the push-rewrite tool, as it completely replaces it AFAICS? We shouldn't establish two new workflows, just one.

mvollmer commented on bots/github-push:
```diff
@@ -0,0 +1,125 @@
+    api = github.GitHub(repo=opts.repo)
+
+    local = git("rev-parse", "HEAD").strip()
+    remote = git("rev-parse", "@{u}").strip()
```
I think we should make github-push work with all the ways we setup our branches.  I can't think of anything fundamental that would be required to make github-push work. 

How did you set up that branch?  Does `git rev-parse @{p}` work in your branch?

I think we can even make github-push work without having set up tracking branches, but it should take advantage of it if there is one.

martinpitt replied:
`@{push}` doesn't work either, same error ("fatal: no upstream configured for branch 'fix-cont-atomic')

How I set it up:
```
git clone https://github.com/cockpit-project/cockpit.git
cd cockpit
git remote add my git@github.com:martinpitt/cockpit.git
git checkout -b somefix
git push my
```

I'm also happy to adjust my workflow; I suppose another common one is to have "origin" be your own cloned project, and have the real upstream repo as other remote name? Would that work?

mvollmer replied:
Now you can give the pull number to github-push and it wont need a upstream tracking branch.

mvollmer commented on bots/github-push:
```diff
@@ -0,0 +1,125 @@
+    if not pull_config:
+        # Find the pull request that matches the tracking branch
+        pulls = api.pulls()
+        for p in pulls:
```
> The API allows you to filter by branch name,

Yeah, but you also need to know the user upfront, no?  I'll experiment with this to figure that out.

martinpitt replied:
How do yo mean? A remote name always uniquely maps to a user, so if I push to a remote like "my" or "cockpituous", .git/config maps that to an URL like git@github.com:martinpitt/cockpit.git which has the user name.

mvollmer commented:
Pull request has been rewritten.
[Diff](https://github.com/cockpit-project/cockpit/compare/23174cf0ab2446f0b7b773bdd9058426fc36f354..d38671832f582a00809f4fa93c436e639388a9cc)

mvollmer commented:
<details ><summary>New commits.</summary>

[Diff](https://github.com/cockpit-project/cockpit/compare/d38671832f582a00809f4fa93c436e639388a9cc..afc53304c0e3e0a67ad7012de67326315cb3f6fc)
```diff
diff --git a/bots/github-push b/bots/github-push
index 363801bef..c09881bf0 100755
--- a/bots/github-push
+++ b/bots/github-push
@@ -7,9 +7,14 @@
 # history of the PR in its comments, regardless of rebases, rewrites,
 # squashed fixups, etc.
 #
-# The tool finds the pull request number by looking through all open
-# pull requests on GitHub.  Once it has been found, the number is
-# cached in the Git config file to speed up subsequent invocations.
+# If you don't have a upstream tracking branch, you need to specify
+# the pull request number as an argument.
+#
+# If you do have a upstream tracking branch, you can omit the pull
+# request number.  The tool will find it by looking through all open
+# pull requests on GitHub and picking the one that matches the
+# tracking branch.  Once it has been found, the number is cached in
+# the Git config file to speed up subsequent invocations.
 
 import subprocess
 import argparse
@@ -35,13 +40,7 @@ def git_config(name):
         return None
     return output.strip()
 
-def main():
-    parser = argparse.ArgumentParser(description='Force push after a rewrite')
-    parser.add_argument('--repo', help="The GitHub repository to work with", default="cockpit-project/cockpit")
-    opts = parser.parse_args()
-
-    api = github.GitHub(repo=opts.repo)
-
+def with_tracking_branch(api):
     local = git("rev-parse", "HEAD").strip()
     remote = git("rev-parse", "@{u}").strip()
 
@@ -121,5 +120,81 @@ def main():
 
     api.post("issues/{0}/comments".format(pull_number), { "body": comment })
 
+def without_tracking_branch(api, pull_number):
+    pull = api.get("pulls/{0}".format(pull_number))
+
+    local = git("rev-parse", "HEAD").strip()
+    remote = pull["head"]["sha"]
+
+    if local == remote:
+        sys.stderr.write("Nothing to push.\n")
+        return 0
+
+    git("fetch", pull["head"]["repo"]["full_name"], pull["head"]["ref"])
+
+    push_dest = pull["head"]["repo"]["full_name"]
+    push_refspecs = local + ":" + pull["head"]["ref"]
+    push_lease = pull["head"]["ref"] + ":" + remote
+
+    diff = git("diff", remote, local)
+    old_commits_link = "https://github.com/" + opts.repo + "/commits/" + remote
+    diff_link = "https://github.com/" + opts.repo + "/compare/" + remote + ".." + local
+
+    def make_comment(title, include_old_commits=False):
+        diff_lines = diff.count("\n")
+        if include_old_commits:
+            first_line = "[Old commits](%s) | [Diff](%s)" % (old_commits_link, diff_link)
+        else:
+            first_line = "[Diff](%s)" % (diff_link)
+        if diff_lines > 0 and diff_lines < 500:
+            diff_open = "open" if diff_lines < 50 else ""
+            return ("<details %s><summary>%s</summary>\n\n%s\n```diff\n%s```\n</details>\n"
+                    % (diff_open, title, first_line, diff))
+        else:
+            return "%s\n%s\n" % (title, first_line)
+
+    git("push", "--force-with-lease=" + push_lease, push_dest, push_refspecs)
+
+    for n in range(100,0,-1):
+        try:
+            if api.get("commits/%s" % local, verbose=False):
+                break
+        except RuntimeError as e:
+            if not "Unprocessable Entity" in str(e) or n <= 1:
+                raise
+            print("(new commits not yet in the GiHub API...please stand by. *beep*)")
+        time.sleep(1)
+
+    if git("merge-base", local, remote).strip() == remote:
+        comment = make_comment("New commits.")
+
+    elif diff == "":
+        old_statuses = api.statuses(remote)
+        copy_count = 0
+        for key in old_statuses:
+            if old_statuses[key]["state"] != "pending":
+                api.post("statuses/" + local, old_statuses[key])
+                copy_count += 1
+
+        comment = make_comment("Commit history has been rewritten.", include_old_commits=True)
+        comment += "\n%s test results have been copied.\n" % copy_count
+    else:
+        comment = make_comment("Pull request has been rewritten.")
+
+    api.post("issues/{0}/comments".format(pull_number), { "body": comment })
+
+def main():
+    parser = argparse.ArgumentParser(description='Force push after a rewrite')
+    parser.add_argument('--repo', help="The GitHub repository to work with", default="cockpit-project/cockpit")
+    parser.add_argument('pull', nargs='?', help="The pull request number to push to")
+    opts = parser.parse_args()
+
+    api = github.GitHub(repo=opts.repo)
+
+    if not opts.pull:
+        return with_tracking_branch(api)
+    else:
+        return without_tracking_branch(api, opts.pull)
+
 if __name__ == '__main__':
     sys.exit(main())
```
</details>

mvollmer commented:
<details ><summary>Pull request has been rewritten.</summary>

[Diff](https://github.com/cockpit-project/cockpit/compare/afc53304c0e3e0a67ad7012de67326315cb3f6fc..9ce5bf91ff9d94440d4510057d1dab39d6aa3663)
```diff
diff --git a/bots/github-push b/bots/github-push
index c09881bf0..b1b66f160 100755
--- a/bots/github-push
+++ b/bots/github-push
@@ -40,42 +40,14 @@ def git_config(name):
         return None
     return output.strip()
 
-def with_tracking_branch(api):
-    local = git("rev-parse", "HEAD").strip()
-    remote = git("rev-parse", "@{u}").strip()
-
+def push_and_comment(api, repo, pull_number, remote, local, push_dest, push_ref):
     if local == remote:
         sys.stderr.write("Nothing to push.\n")
         return 0
 
-    ( tracking_remote, tracking_ref ) = git("rev-parse", "--abbrev-ref", "--symbolic-full-name", "@{u}").strip().split("/")
-    tracking_url = git("remote", "get-url", tracking_remote).strip()
-
-    push_dest = tracking_remote
-    push_refspecs = local + ":" + tracking_ref
-    push_lease = tracking_ref + ":" + remote
-
-    pull_number = None
-    pull_config = git_config("pull.%s.%s" % (tracking_remote, tracking_ref))
-    if not pull_config:
-        # Find the pull request that matches the tracking branch
-        pulls = api.pulls()
-        for p in pulls:
-            ref = p["head"]["ref"]
-            url = "git@github.com:%s.git" % p["head"]["repo"]["full_name"]
-            if ref == tracking_ref and url == tracking_url:
-                pull_number = p["number"]
-                break
-        if not pull_number:
-            sys.stderr.write("No pull request found for %s/%s\n" % (tracking_remote, tracking_ref))
-            return 1
-        git("config", "pull.%s.%s" % (tracking_remote, tracking_ref), str(pull_number))
-    else:
-        pull_number=int(pull_config)
-
-    diff = git("diff",remote, local)
-    old_commits_link = "https://github.com/" + opts.repo + "/commits/" + remote
-    diff_link = "https://github.com/" + opts.repo + "/compare/" + remote + ".." + local
+    diff = git("diff", remote, local)
+    old_commits_link = "https://github.com/" + repo + "/commits/" + remote
+    diff_link = "https://github.com/" + repo + "/compare/" + remote + ".." + local
 
     def make_comment(title, include_old_commits=False):
         diff_lines = diff.count("\n")
@@ -90,7 +62,7 @@ def with_tracking_branch(api):
         else:
             return "%s\n%s\n" % (title, first_line)
 
-    git("push", "--force-with-lease=" + push_lease, push_dest, push_refspecs)
+    git("push", "--force-with-lease=%s:%s" % (push_ref, remote), push_dest, "%s:%s" % (local, push_ref))
 
     for n in range(100,0,-1):
         try:
@@ -109,9 +81,8 @@ def with_tracking_branch(api):
         old_statuses = api.statuses(remote)
         copy_count = 0
         for key in old_statuses:
-            if old_statuses[key]["state"] != "pending":
-                api.post("statuses/" + local, old_statuses[key])
-                copy_count += 1
+            api.post("statuses/" + local, old_statuses[key])
+            copy_count += 1
 
         comment = make_comment("Commit history has been rewritten.", include_old_commits=True)
         comment += "\n%s test results have been copied.\n" % copy_count
@@ -120,68 +91,50 @@ def with_tracking_branch(api):
 
     api.post("issues/{0}/comments".format(pull_number), { "body": comment })
 
-def without_tracking_branch(api, pull_number):
-    pull = api.get("pulls/{0}".format(pull_number))
-
+def with_tracking_branch(api, repo):
     local = git("rev-parse", "HEAD").strip()
-    remote = pull["head"]["sha"]
-
-    if local == remote:
-        sys.stderr.write("Nothing to push.\n")
-        return 0
-
-    git("fetch", pull["head"]["repo"]["full_name"], pull["head"]["ref"])
-
-    push_dest = pull["head"]["repo"]["full_name"]
-    push_refspecs = local + ":" + pull["head"]["ref"]
-    push_lease = pull["head"]["ref"] + ":" + remote
-
-    diff = git("diff", remote, local)
-    old_commits_link = "https://github.com/" + opts.repo + "/commits/" + remote
-    diff_link = "https://github.com/" + opts.repo + "/compare/" + remote + ".." + local
-
-    def make_comment(title, include_old_commits=False):
-        diff_lines = diff.count("\n")
-        if include_old_commits:
-            first_line = "[Old commits](%s) | [Diff](%s)" % (old_commits_link, diff_link)
-        else:
-            first_line = "[Diff](%s)" % (diff_link)
-        if diff_lines > 0 and diff_lines < 500:
-            diff_open = "open" if diff_lines < 50 else ""
-            return ("<details %s><summary>%s</summary>\n\n%s\n```diff\n%s```\n</details>\n"
-                    % (diff_open, title, first_line, diff))
-        else:
-            return "%s\n%s\n" % (title, first_line)
+    remote = git("rev-parse", "@{u}").strip()
 
-    git("push", "--force-with-lease=" + push_lease, push_dest, push_refspecs)
+    ( tracking_remote, tracking_ref ) = git("rev-parse", "--abbrev-ref", "--symbolic-full-name", "@{u}").strip().split("/")
+    tracking_url = git("remote", "get-url", tracking_remote).strip()
 
-    for n in range(100,0,-1):
-        try:
-            if api.get("commits/%s" % local, verbose=False):
+    pull_number = None
+    pull_config = git_config("pull.%s.%s" % (tracking_remote, tracking_ref))
+    if not pull_config:
+        # Find the pull request that matches the tracking branch
+        pulls = api.pulls()
+        for p in pulls:
+            ref = p["head"]["ref"]
+            url = "git@github.com:%s.git" % p["head"]["repo"]["full_name"]
+            if ref == tracking_ref and url == tracking_url:
+                pull_number = p["number"]
                 break
-        except RuntimeError as e:
-            if not "Unprocessable Entity" in str(e) or n <= 1:
-                raise
-            print("(new commits not yet in the GiHub API...please stand by. *beep*)")
-        time.sleep(1)
+        if not pull_number:
+            sys.stderr.write("No pull request found for %s/%s\n" % (tracking_remote, tracking_ref))
+            return 1
+        git("config", "pull.%s.%s" % (tracking_remote, tracking_ref), str(pull_number))
+    else:
+        pull_number=int(pull_config)
 
-    if git("merge-base", local, remote).strip() == remote:
-        comment = make_comment("New commits.")
+    push_and_comment(api, repo, pull_number,
+                     local=local, remote=remote,
+                     push_dest=tracking_remote, push_ref=tracking_ref)
 
-    elif diff == "":
-        old_statuses = api.statuses(remote)
-        copy_count = 0
-        for key in old_statuses:
-            if old_statuses[key]["state"] != "pending":
-                api.post("statuses/" + local, old_statuses[key])
-                copy_count += 1
+def without_tracking_branch(api, repo, pull_number):
+    pull = api.get("pulls/{0}".format(pull_number))
 
-        comment = make_comment("Commit history has been rewritten.", include_old_commits=True)
-        comment += "\n%s test results have been copied.\n" % copy_count
-    else:
-        comment = make_comment("Pull request has been rewritten.")
+    local = git("rev-parse", "HEAD").strip()
+    remote = pull["head"]["sha"]
 
-    api.post("issues/{0}/comments".format(pull_number), { "body": comment })
+    # XXX - If we don't have the remote sha already in our local
+    # repository, then we should probably fail here....
+    #
+    git("fetch", "git@github.com:%s.git" % pull["head"]["repo"]["full_name"], pull["head"]["ref"])
+
+    push_and_comment(api, repo, pull_number,
+                     local=local, remote=remote,
+                     push_dest="git@github.com:%s.git" % pull["head"]["repo"]["full_name"],
+                     push_ref=pull["head"]["ref"])
 
 def main():
     parser = argparse.ArgumentParser(description='Force push after a rewrite')
@@ -192,9 +145,9 @@ def main():
     api = github.GitHub(repo=opts.repo)
 
     if not opts.pull:
-        return with_tracking_branch(api)
+        return with_tracking_branch(api, opts.repo)
     else:
-        return without_tracking_branch(api, opts.pull)
+        return without_tracking_branch(api, opts.repo, opts.pull)
 
 if __name__ == '__main__':
     sys.exit(main())
```
</details>

mvollmer commented:
<details><summary>New commits.</summary>

[Diff](https://github.com/cockpit-project/cockpit/compare/3ea76a7abb44a0274485c857e5e5d41624f172a0..f8e083725d90ab40936383022feffc553ab903c5)
```diff
diff --git a/bots/github-push b/bots/github-push
index 80f7a1dc4..bfae6227d 100755
--- a/bots/github-push
+++ b/bots/github-push
@@ -51,14 +51,15 @@ def push_and_comment(api, repo, pull_number, remote, local, push_dest, push_ref)
 
     def make_comment(title, include_old_commits=False):
         diff_lines = diff.count("\n")
+        first_line_parts = [ ]
         if include_old_commits:
-            first_line = "[Old commits](%s) | [Diff](%s)" % (old_commits_link, diff_link)
-        else:
-            first_line = "[Diff](%s)" % (diff_link)
+            first_line_parts.append("[Old commits](%s)" % old_commits_link)
+        if diff_lines > 0:
+            first_line_parts.append("[Diff](%s)" % (diff_link))
+        first_line = " | ".join(first_line_parts)
         if diff_lines > 0 and diff_lines < 500:
-            diff_open = "open" if diff_lines < 50 else ""
-            return ("<details %s><summary>%s</summary>\n\n%s\n```diff\n%s```\n</details>\n"
-                    % (diff_open, title, first_line, diff))
+            return ("<details><summary>%s</summary>\n\n%s\n```diff\n%s```\n</details>\n"
+                    % (title, first_line, diff))
         else:
             return "%s\n%s\n" % (title, first_line)
 
```
</details>

mvollmer commented:
Commit history has been rewritten.
[Old commits](https://github.com/cockpit-project/cockpit/commits/f8e083725d90ab40936383022feffc553ab903c5)

29 test results have been copied.

martinpitt commented on bots/github-push:
```diff
@@ -0,0 +1,154 @@
+        if diff_lines > 0:
+            first_line_parts.append("[Diff](%s)" % (diff_link))
+        first_line = " | ".join(first_line_parts)
+        if diff_lines > 0 and diff_lines < 500:
```
IMHO 500 is quite an enormous comment; the /compare link should suffice for most use cases and doesn't clobber the comments then. But this is not a veto, we can try this and change this in the future to our liking.

martinpitt commented on bots/github-push:
```diff
@@ -0,0 +1,154 @@
+
+    diff = git("diff", remote, local)
+    old_commits_link = "https://github.com/" + repo + "/commits/" + remote
+    diff_link = "https://github.com/" + repo + "/compare/" + remote + ".." + local
```
The existence of this github API is a really nice little gem! So the most important thing on rewriting push requests is to record all the HEADs somewhere.

martinpitt commented on bots/github-push:
```diff
@@ -0,0 +1,154 @@
+        copy_count = 0
+        for key in old_statuses:
+            api.post("statuses/" + local, old_statuses[key])
+            copy_count += 1
```
`copy_count` could just be replaced with `len(old_statuses)`. But nitpicking..

martinpitt commented on bots/github-push:
```diff
@@ -0,0 +1,154 @@
+    remote = pull["head"]["sha"]
+
+    # XXX - If we don't have the remote sha already in our local
+    # repository, then we should probably fail here....
```
Is this something which should be addressed before landing this?

martinpitt commented:
Meh, it bitrotted a bit, now needs rebasing. @mvollmer, can we settle the question about how severe the "TODO" is, and land this?

mvollmer commented:
Regarding the TODO.

If you have never fetched the current tip of the remote branch, the tool can not produce the diff and do other parts of its work. So currently the tool makes sure that the remote branch tip is in the local repo.

However, if you have never fetched that tip, you are very likely going to overwrite work that you have never seen. So it would be better to just fail and require the user to do the fetch.

This only applies in the "no tracking branch" case.

martinpitt commented:
@mvollmer: See the previous comment - do you still want to go through with this, or just keep the current push-rewrite?

martinpitt commented:
I force-pushed a rebase to unblock our CI.

marusak commented:
Needs to rebase to master since #12367 changed tests names

mvollmer commented:
I am afraid I lost interest in this...
